Compute values of basic shape rect() and xywh() as the equivalent inset()

darinadler opened the pull request:
#### 534a722e7081523b73c20634c2932ab927c499a9
<pre>
Compute values of basic shape rect() and xywh() as the equivalent inset()
<a href="https://bugs.webkit.org/show_bug.cgi?id=266513">https://bugs.webkit.org/show_bug.cgi?id=266513</a>
<a href="https://rdar.apple.com/problem/119739406">rdar://problem/119739406</a>

Reviewed by Tim Nguyen.

CSS Shapes called for this behavior but we did not implement it yet.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/animations/clip-path-interpolation-xywh-rect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-path-interpolation-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
More PASS, less FAIL.

* Source/WebCore/css/BasicShapeFunctions.cpp:
(WebCore::valueForBasicShape): Added code that computes the equivalent inset as specified.
Use the CSSCalcValue machinery as needed.

Canonical link: <a href="https://commits.webkit.org/272180@main">https://commits.webkit.org/272180@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/977dfbc87305f79de3aae6ec121f53540a459f64

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30765 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9443 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32432 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33260 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27798 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31505 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11748 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6689 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27678 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31078 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7922 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27521 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6795 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6984 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27397 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34597 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27995 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27879 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33129 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6985 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5074 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30936 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8702 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7291 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7702 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7541 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/BasicShapeFunctions.cpp:
```diff
@@ -80,12 +83,39 @@ static SVGPathByteStream copySVGPathByteStream(const SVGPathByteStream& source,
 {
     auto createValue = [&](const Length& length) {
-        return CSSPrimitiveValue::create(length, style);
+        return CSSPrimitiveValue::create(length.isAuto() ? Length(0, LengthType::Percent) : length, style);
```
From https://drafts.csswg.org/css-shapes-1/#basic-shape-computed-values :

> An auto value makes the edge of the box coincide with the corresponding edge of the [reference box](https://drafts.csswg.org/css-shapes-1/#basic-shape-reference-box): it’s equivalent to 0% as the first (top) or fourth (left) value, and equivalent to 100% as the second (right) or third (bottom) value.

Is this conversion correct? Maybe converting to 100% in the relevant cases would fix the interpolation tests?

darinadler replied:
It’s correct. There are no `100` cases, because all the `auto` are in `rect` and we are always converting to `inset` and so it’s always `0` once it’s a value in the `inset`. Earlier ways I wrote the code had some `100` cases because I wrote it based on the above text, but as I refined and refactored the code, it led to it just being this. I’ve looked at remaining failing interpolation tests and the issues are the interpolation not the values.

darinadler replied:
Oh, actually, I thought of another way to explain the refactoring. Rather than have a separate case of `100` in here, I put a special case for `auto` into `createReflectedValue`. If I had not done that, then I would have had to pass in a value of `0` or `100`, and it would have computed `100% - 100%` and ended up `0%` anyway.

nt1m reviewed (approved):
Looks good aside from one comment, thanks!

darinadler committed:
Compute values of basic shape rect() and xywh() as the equivalent inset()
https://bugs.webkit.org/show_bug.cgi?id=266513
rdar://problem/119739406

Reviewed by Tim Nguyen.

CSS Shapes called for this behavior but we did not implement it yet.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/animations/clip-path-interpolation-xywh-rect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-path-interpolation-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
More PASS, less FAIL.

* Source/WebCore/css/BasicShapeFunctions.cpp:
(WebCore::valueForBasicShape): Added code that computes the equivalent inset as specified.
Use the CSSCalcValue machinery as needed.

Canonical link: https://commits.webkit.org/272180@main
